### DISCUSSION TITLE
Batches and generation

### DISCUSSION THREAD

**Discussion by erikfransson** [brucefan1983/GPUMD#1137] [Announcements]
The batching  and generations during trained currently seem to work as (with 2 batches)
```
Start gen 1
load batch 1 --> update parameters
Start gen 2
load batch 2 --> update parameters
Start gen 3
load batch 1 --> update parameters
...
```
and so on.
This mean "one generation" is not a complete pass through the entire training dataset (as one epoch would be). 

1. It  means loss and RMSE are only computed for the current batch of structures and not over the total training set, making Loss and RMSE oscillate during training (even when parameters are not changing between generations) which can be quite confusing for a user (me :))

<img width="400"  alt="Image" src="https://github.com/user-attachments/assets/e617d5bc-a317-4bca-9a84-97d23b642a49" />

2. Additionally, it means that if I run 100k generations with batch=10 or batch=1000 Im doing much "more training" in the latter case, also in my case without knowing it.

3. Lastly, I think when doing mini-batches its common to randomly re-shuffle the batches each "step/epoch" atleast in gradient descent methods to maybe avoid artificial/correlated behaviors, dont know how relevant this is for SNES.

---
I think defining a training step similar to an epoch, like something below, would be more intuitive for users and reporting loss/rmse for entire training set would be more understandable.

```
Start step 1
load batch 1 --> update parameters
load batch 2 --> update parameters
...
Start step 2
load batch 1 --> update parameters
load batch 2 --> update parameters
```

**Comment by brucefan1983** [comment]
> Lastly, I think when doing mini-batches its common to randomly re-shuffle the batches each "step/epoch" atleast in gradient descent methods to maybe avoid artificial/correlated behaviors, dont know how relevant this is for SNES.

I tried re-shuffling the batches for each epoch, but it turned out that SNES does not like this.

**Comment by brucefan1983** [comment]
Generally, SNES favors full-batch training, but mini-batch training is OK during the active learning process.

**Comment by brucefan1983** [comment]
This is better suited to the `Discussions`.
